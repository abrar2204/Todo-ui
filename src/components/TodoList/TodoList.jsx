import React, {useEffect, useState} from 'react';
import axios from "axios";

const TodoList = () => {
  const [todos,setTodos]=useState([]);

  useEffect(()=>{
    axios.get('/api/todo').then( res =>{
      if(res.data.success){
        setTodos(res.data.success);
      }
    })
  },[]);

  const deleteTodo = (todoId)=>{
    axios.delete(`/api/todo/${todoId}`).then(res=>{
      if(res.data.success){
        setTodos(prevTodos=>prevTodos.filter(({id})=>id!==todoId));
      }
    })
  }

  return(
    <div>
      <h1>Welcome to your TodoList</h1>
      {
        todos.map(todo =>
          <div data-testid={`todo-${todo.id}`} key={todo.id}>
            <h3>{todo.title}</h3>
            <button data-testid={`delete-button-${todo.id}`} onClick={()=>deleteTodo(todo.id)}>Delete</button>
          </div>
        )
      }
    </div>
  );
};

export default TodoList;