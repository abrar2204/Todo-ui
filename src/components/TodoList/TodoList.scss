@import "../../styles/colors";

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;

  h1 {
    font-size: 5rem;
    font-weight: 600;
    margin: 3rem;
  }

  .show-form-button {
    background-color: $primary-dark-color;
    outline: none;
    border: none;
    height: 4.8rem;
    border-radius: 5px;

    img {
      transition: transform .5s ease-in;
    }

    &.open {
      & > img {
        transform: rotateZ(135deg);
      }
    }
  }
}

.form {
  display: flex;
  flex-direction: column;
  width: 50rem;
  max-height: 0;
  overflow: hidden;
  transition: max-height .5s ease-in;

  &.open {
    max-height: 225px;
  }

  &-control {
    display: flex;
    flex-direction: column;
    margin: 0.75rem 0;

    input {
      margin: 0.5rem 0;
      background-color: $primary-dark-color;
      border: none;
      outline: none;
      padding: 1rem 1.5rem;
      border-radius: 5px;
      color: $primary-font-color;

      &:focus {
        outline: solid 1px $secondary-font-color;
      }
    }
  }

  button {
    padding: 1rem;
    margin: 0.7rem 0;
    border-radius: 5px;
    background-color: $background-color;
    color: $primary-font-color;
    border: 2px solid $primary-font-color;
    outline: none;

    &:focus {
      outline: solid 1px $secondary-font-color;
    }

  }
}

.todos {
  padding: 2rem 0;

  .todo {
    display: flex;
    align-items: flex-start;
    justify-content: flex-start;
    padding: 2rem;
    margin: 2rem 0;
    width: 80rem;
    border-bottom: 3px solid #343A3F;

    &-checkbox {
      flex-basis: 7rem;
      position: relative;

      .checkmark {
        position: absolute;
        top: 0;
        left: 0;
        height: 3.2rem;
        width: 3.2rem;
        background-color: #484F58;
        border: 1px solid $primary-dark-color;
        border-radius: 5px;

        &:hover {
          cursor: pointer;
        }

        &:after {
          content: "";
          display: none;
          background: url(https://img.icons8.com/ios/25/FFFFFF/checkmark--v1.png);
          width: 2.5rem;
          height: 2.5rem;
          position: absolute;
          left: 2px;
          top: 1px;
        }

        &.checked {
          &:after {
            display: block;
          }
        }
      }
    }

    &-data {
      flex-basis: 63rem;

      h3 {
        font-size: 2.2rem;
        font-weight: 600;
      }

      p {
        color: $secondary-font-color;
      }
    }

    &-button-group {
      flex-basis: 10rem;

      button {
        padding: 5px;
        margin: 10px 5px;
        border-radius: 5px;
        outline: none;
        border: 0;
        cursor: pointer;
        background-color: #484F58;
      }
    }
  }
}