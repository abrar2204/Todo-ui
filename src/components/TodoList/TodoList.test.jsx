import React from 'react';
import "@testing-library/jest-dom";
import TodoList from "./TodoList";
import {render, screen, waitFor} from '@testing-library/react';
import axios from 'axios';
import userEvent from "@testing-library/user-event";

const getAllTodosResponse = {
  "success": [
    {
      "id": 1,
      "title": "Download Game",
      "description": "Install Monster Hunter Rise Sunbreak",
      "completed": true,
      "createdAt": "2022-06-30"
    },
    {
      "id": 2,
      "title": "Pay Bills",
      "description": "Pay Internet and Electricity Bill",
      "completed": false,
      "createdAt": "2020-06-01"
    },
    {
      "id": 3,
      "title": "Fill Petrol",
      "description": "Fill Petrol in Bike",
      "completed": true,
      "createdAt": "2020-06-27"
    },
  ],
  "error": null
}

jest.mock('axios');

describe("TodoList",()=>{

  it("should render a todo list",()=>{
    const {asFragment} = render(<TodoList/>);

    expect(screen.getByText("Welcome to your TodoList")).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  })

  it("should render a list of todos",async ()=>{
    axios.get.mockResolvedValueOnce({ data: getAllTodosResponse });
    const {asFragment} = render(<TodoList/>);

    expect(screen.getByText("Welcome to your TodoList")).toBeInTheDocument();

    await waitFor(()=>{
      expect(screen.queryByTestId("todo-1")).toBeInTheDocument();
      expect(screen.queryByTestId("todo-2")).toBeInTheDocument();
      expect(screen.queryByTestId("todo-3")).toBeInTheDocument();
    })

    expect(asFragment).toMatchSnapshot();
  })

  it("should delete a todo",async()=>{
    const user = userEvent.setup();
    axios.get.mockResolvedValueOnce({ data: getAllTodosResponse });
    axios.delete.mockResolvedValueOnce({data: {
        "success": "Successfully deleted 1 todo",
        "error": null
    }})
    const {asFragment} = render(<TodoList/>);

    expect(screen.getByText("Welcome to your TodoList")).toBeInTheDocument();
    await waitFor(()=>{
      expect(screen.getByTestId("todo-1")).toBeInTheDocument();
    })
    await user.click(screen.getByTestId('delete-button-1'));
    await waitFor(()=>{
      expect(screen.queryByTestId("todo-1")).not.toBeInTheDocument();
    })

    expect(asFragment).toMatchSnapshot();
  })


})