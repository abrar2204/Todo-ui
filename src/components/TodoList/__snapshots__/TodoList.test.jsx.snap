// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should delete a todo 1`] = `[Function]`;

exports[`TodoList should render a list of todos 1`] = `[Function]`;

exports[`TodoList should render a todo list 1`] = `[Function]`;
